### DISCUSSION TITLE
alist快把物理内存占完了

### DISCUSSION THREAD

**Discussion by leetomlee123** [AlistGo/alist#5185] [General]
主要调用alist接口 上传文件 内存慢慢涨上来了 使用了"As-Task"="true"
![image](https://github.com/alist-org/alist/assets/19498940/6d3267c2-93f5-4c89-8831-2d55fbdfc941)

**Comment by anwen-anyi** [comment]
上传好了就会释放吧~
1. stream直接把文件二进制内容放进body中，这样后端可以使用很少的内存，但是浏览器可能会限制大小因为要一次性把文件读进内存。
2. form是把文件包裹成formdata上传，后端使用的内存会多一些，但是不限制大小。

**Reply by xhofe** [reply]
1. ~~但是浏览器可能会限制大小因为要一次性把文件读进内存~~ 已经不会这样了

**Reply by anwen-anyi** [reply]
OK 这就改过来。
现在是：`stream直接把文件二进制内容放进body中，这样后端可以使用很少的内存`这样？

**Reply by xhofe** [reply]
文档中已经改过了 https://alist.nn.ci/zh/faq/why.html#%E4%B8%A4%E4%B8%AA%E4%B8%8A%E4%BC%A0%E6%9C%89%E4%BB%80%E4%B9%88%E5%8C%BA%E5%88%AB
不知道还有没有地方有这个

**Reply by leetomlee123** [reply]
我是很多个1 2m的音频文件  内存慢慢涨上来的 目测有内存泄漏的问题

**Comment by xhofe** [comment]
说一下详细的复现步骤

**Comment by leetomlee123** [comment]
`
	formURL := "http://" + host + "/api/fs/form"
	filename := path // File to be uploaded
	file, err := os.Open(filename)
	if err != nil {
		logrus.Println("Failed to open file:", err)
		return
	}
	defer file.Close()

	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	part, err := writer.CreateFormFile("file", filename)
	if err != nil {
		logrus.Println("Failed to create form file:", err)
		return
	}
	_, err = io.Copy(part, file)
	if err != nil {
		logrus.Println("Failed to copy file to form:", err)
		return
	}
	err = writer.Close()
	if err != nil {
		logrus.Println("Failed to close writer:", err)
		return
	}

	formReq, err := http.NewRequest("PUT", formURL, body)
	if err != nil {
		logrus.Println("Failed to create form request:", err)
		return
	}
	formReq.Header.Set("Authorization", token)
	formReq.Header.Set("As-Task", "true")
	formReq.Header.Set("Content-Type", writer.FormDataContentType())
	formReq.Header.Set("file-path", "/root/"+bookId+"/"+uid+"/"+filename)

	formClient := &http.Client{}
	formRes, err := formClient.Do(formReq)
	if err != nil {
		logrus.Println("Failed to send form request:", err)
		return
	}
	defer formRes.Body.Close()

	if formRes.StatusCode == http.StatusOK {
		logrus.Println("Upload success")
	}`   今天就变动了formReq.Header.Set("As-Task", "true") 就内存出问题了

**Reply by xhofe** [reply]
所以只是 form+as task 上传文件至本地？

**Reply by leetomlee123** [reply]
下载文件到本地 然后再传到alist  alist响应成功 就删除本地文件  ，用了as task 就会爆内存 ，无论表单还是流式上传

**Reply by xhofe** [reply]
form 可以复现内存泄漏，已提交修复。
stream 无法复现，上传内存均正常。
